File: design_patterns/src/builder/c1.rs
/*
建造者模式
 */

// todo: 1. 目标类 Person
#[derive(Clone, Debug)]
struct Person {
  name:     String,     // 必须
  age:      u8,         // 必须
  job:      Option<String>,
}

impl Person {
  fn new(name: String, age: u8) -> Person {
    Person {
      // name:     Default::default(),
      name:     name,
      // age:      Default::default(),
      age:      age,
      job:      None,
    }
  }

  fn set_job(&mut self, job: Option<String>) {
    self.job = job
  }
}

// ----------------------------------------------------
// todo: 2. 抽象构建者类
trait Builder {
  // fn  build_name(&mut self);
  // fn  build_age(&mut self);
  fn  build_job(&mut self);
  fn  build(&mut self) -> Person;
}

// ----------------------------------------------------
// todo: 3. 实体构建者类
struct AliceBuilder {
  obj:    Person,        // 实例化
}

impl AliceBuilder {
  fn new() -> AliceBuilder {
    AliceBuilder {
      obj:    Person::new("Alice".to_string(), 12),
    }
  }
}

impl Builder for AliceBuilder {
  fn build_job(&mut self) {
    self.obj.set_job(Some("Student".to_string()))
  }

  fn build(&mut self) -> Person {
    self.obj.clone()
  }
}

// todo: 4. 主导类 Director
struct PersonDirector {
  build:  Box<dyn Builder>,   // todo: 指向抽象构建类
}

impl PersonDirector {
  fn new(b :Box<dyn Builder>) -> PersonDirector {
    PersonDirector {
      build: b,
    }
  }

  fn build(&mut self) -> Person {
    self.build.build_job();     // todo: 这里只有方法调用， 不指定参数
    self.build.build()
  }
}

fn main() {
  // todo:  from struct abstract to trait
  let alice_builder = Box::new(AliceBuilder::new()) as Box<dyn Builder>;

  let mut director = PersonDirector::new(alice_builder);
  let alice = director.build();

  println!("alice: {:?}", alice);
}

File: design_patterns/src/builder/go/c1.go
package main

import "log"

/*
建造者模式
 */

// todo: 1. 目标类
type Person struct {
  name    string      // 必须
  age     uint8       // 必须
  job     string
}

func NewPerson(name string, age uint8) Person {
  return Person{
    name: name,
    age:  age,
  }
}

func (p *Person) setJob(job string) {
  p.job = job
}

// -----------------------------------------------------
// todo: 2. 抽象构建者类
type Builder interface {
  buildJob()
  build() Person
}

// ----------------------------------------------------
// todo: 3. 实体构建者类
type AliceBuilder struct {
  obj     Person
}

func NewAliceBuilder() AliceBuilder {
  return AliceBuilder{
    obj:  NewPerson("Alice", 12),
  }
}

func (alice *AliceBuilder) buildJob() {
  alice.obj.setJob("Student")
}

func (alice *AliceBuilder) build() Person {
  return alice.obj
}

// todo: 4. 主导类 Director
type PersonDirector struct {
  build     Builder   // todo: 指向抽象构建类
}

func NewPersonDirector(b Builder) PersonDirector {
  return PersonDirector{
    build:  b,
  }
}

func (p *PersonDirector) buildx() Person {
  p.build.buildJob()        // todo: 这里只有方法调用， 不指定参数
  return p.build.build()
}

func main() {
  aliceBuilder := NewAliceBuilder()     // 1. 建立一个builder
  director := NewPersonDirector(&aliceBuilder)      // 2. 建立一个director， load builder参数
  alice := director.buildx()            // 3. 通过 director 来 build alice
  log.Printf("alice ---------------- %+v", alice)
}






